update to show Zero values as in the ticket SCT-439

hnadiminti-equinix opened the pull request:
Desc:
The code has been modified to show zero values as in the ticket([SCT-439](https://packet.atlassian.net/browse/SCT-439)).
If a configuration is not present in the default config then that value won't come from the template as it checks for null values.

```
plugins:
    trivy:
       exit-code: 0
       timeout : 5m0s
       severity: HIGH
       ignore-unfixed: true
       security-checks: vuln,config
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
```

```
go run . trivy
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.2:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
```

```
go run . trivy --exit-code=0 --ignore-unfixed=false
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.2:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH"
          ignore-unfixed: false
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""

go run . trivy --exit-code=42 --ignore-unfixed=false
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.2:
          exit-code: 42
          timeout : "5m0s"
          severity: "HIGH"
          ignore-unfixed: false
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
```

prezha commented on conf.yaml:
```diff
@@ -2,11 +2,12 @@ plugins:
   trivy:
     exit-code: 0
     timeout: 5m0s
+    timeout: 5m0s
```
`timeout` is duplicated?

hnadiminti-equinix replied:
addressed.

prezha commented on generator/common-pipeline_test.go:
```diff
@@ -15,17 +15,22 @@ type testCase struct {
   - command: ls
     plugins:
-      - equinixmetal-buildkite/trivy#:
+      - equinixmetal-buildkite/trivy#<nil>:
```
is this test ok: trivy version should never be nil - ie, GetLatestTrivyPluginTag() would run if it's not provided?

prezha commented on templates/trivy-step.tmpl:
```diff
@@ -7,7 +7,7 @@ steps:
           {{- end }}
           {{- if .TPConfig.Timeout}}
-          timeout : {{ .TPConfig.Timeout }}
+          timeout : "{{ .TPConfig.Timeout }}"
```
nit: timeout should be of time.Duration type and then should not require `"` here, but we change that later

hnadiminti-equinix replied:
Trivy accepts  "Timeout"  value as a string.

prezha reviewed (changes requested):
@hnadiminti-equinix i've left a few minor comments

please also have a look and test - it's segfaulting now:

```
$ git status
On branch SCT-439
Your branch is up to date with 'origin/SCT-439'.

nothing to commit, working tree clean
```

```
$ go run . trivy
{"level":"debug","msg":"config path:conf.yaml","time":"2023-09-22T17:01:09+01:00"}
{"level":"debug","msg":"error while loading the configuration file. Loading the defaults","time":"2023-09-22T17:01:09+01:00"}
{"level":"debug","msg":"trivy plugin configuration not found in the config file or wrong config file. Proceeding with defaults from command line flags.","time":"2023-09-22T17:01:09+01:00"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1046f27f4]

goroutine 1 [running]:
dynamic-buildkite-template/cmd.setFromIntFlag(0x0, 0x100000005?, {0x1046f4b89?, 0x1?}, 0x1?)
	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/util.go:40 +0xa4
dynamic-buildkite-template/cmd.CreateGenerator(0x0?, {0x0?, 0x0?, 0x0?})
	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/trivy.go:61 +0x12c
dynamic-buildkite-template/cmd.glob..func2(0x1400020c200?, {0x104b3c7a0?, 0x4?, 0x1046f36d0?})
	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/trivy.go:25 +0x20
github.com/spf13/cobra.(*Command).execute(0x104b026e0, {0x104b3c7a0, 0x0, 0x0})
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x640
github.com/spf13/cobra.(*Command).ExecuteC(0x104b02400)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x320
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
dynamic-buildkite-template/cmd.Execute(...)
	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/generate.go:50
main.main()
	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/main.go:23 +0x28
exit status 2
```
i think that's because it's switched to pointers, and they need to be initialised - ie, non-nil, before value assignment is attempted

hnadiminti-equinix commented:
> @hnadiminti-equinix i've left a few minor comments
> 
> please also have a look and test - it's segfaulting now:
> 
> ```
> $ git status
> On branch SCT-439
> Your branch is up to date with 'origin/SCT-439'.
> 
> nothing to commit, working tree clean
> ```
> 
> ```
> $ go run . trivy
> {"level":"debug","msg":"config path:conf.yaml","time":"2023-09-22T17:01:09+01:00"}
> {"level":"debug","msg":"error while loading the configuration file. Loading the defaults","time":"2023-09-22T17:01:09+01:00"}
> {"level":"debug","msg":"trivy plugin configuration not found in the config file or wrong config file. Proceeding with defaults from command line flags.","time":"2023-09-22T17:01:09+01:00"}
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1046f27f4]
> 
> goroutine 1 [running]:
> dynamic-buildkite-template/cmd.setFromIntFlag(0x0, 0x100000005?, {0x1046f4b89?, 0x1?}, 0x1?)
> 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/util.go:40 +0xa4
> dynamic-buildkite-template/cmd.CreateGenerator(0x0?, {0x0?, 0x0?, 0x0?})
> 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/trivy.go:61 +0x12c
> dynamic-buildkite-template/cmd.glob..func2(0x1400020c200?, {0x104b3c7a0?, 0x4?, 0x1046f36d0?})
> 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/trivy.go:25 +0x20
> github.com/spf13/cobra.(*Command).execute(0x104b026e0, {0x104b3c7a0, 0x0, 0x0})
> 	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x640
> github.com/spf13/cobra.(*Command).ExecuteC(0x104b02400)
> 	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x320
> github.com/spf13/cobra.(*Command).Execute(...)
> 	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
> dynamic-buildkite-template/cmd.Execute(...)
> 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/generate.go:50
> main.main()
> 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/main.go:23 +0x28
> exit status 2
> ```
> 
> i think that's because it's switched to pointers, and they need to be initialised - ie, non-nil, before value assignment is attempted

I'm going to add nil check as suggested ...WIP, Thank you @prezha

hnadiminti-equinix commented:
> > @hnadiminti-equinix i've left a few minor comments
> > please also have a look and test - it's segfaulting now:
> > ```
> > $ git status
> > On branch SCT-439
> > Your branch is up to date with 'origin/SCT-439'.
> > 
> > nothing to commit, working tree clean
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > $ go run . trivy
> > {"level":"debug","msg":"config path:conf.yaml","time":"2023-09-22T17:01:09+01:00"}
> > {"level":"debug","msg":"error while loading the configuration file. Loading the defaults","time":"2023-09-22T17:01:09+01:00"}
> > {"level":"debug","msg":"trivy plugin configuration not found in the config file or wrong config file. Proceeding with defaults from command line flags.","time":"2023-09-22T17:01:09+01:00"}
> > panic: runtime error: invalid memory address or nil pointer dereference
> > [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1046f27f4]
> > 
> > goroutine 1 [running]:
> > dynamic-buildkite-template/cmd.setFromIntFlag(0x0, 0x100000005?, {0x1046f4b89?, 0x1?}, 0x1?)
> > 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/util.go:40 +0xa4
> > dynamic-buildkite-template/cmd.CreateGenerator(0x0?, {0x0?, 0x0?, 0x0?})
> > 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/trivy.go:61 +0x12c
> > dynamic-buildkite-template/cmd.glob..func2(0x1400020c200?, {0x104b3c7a0?, 0x4?, 0x1046f36d0?})
> > 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/trivy.go:25 +0x20
> > github.com/spf13/cobra.(*Command).execute(0x104b026e0, {0x104b3c7a0, 0x0, 0x0})
> > 	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x640
> > github.com/spf13/cobra.(*Command).ExecuteC(0x104b02400)
> > 	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x320
> > github.com/spf13/cobra.(*Command).Execute(...)
> > 	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
> > dynamic-buildkite-template/cmd.Execute(...)
> > 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/cmd/generate.go:50
> > main.main()
> > 	/go/src/github.com/equinixmetal-buildkite/dynamic-buildkite-template/main.go:23 +0x28
> > exit status 2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i think that's because it's switched to pointers, and they need to be initialised - ie, non-nil, before value assignment is attempted
> 
> I'm going to add nil check as suggested ...WIP, Thank you @prezha

I shifted to pointers for members of ```TrivyPluginConfig``` because we wanted to print zero values. In addition we want to check if a parameter/flag is not defined or has zero values. The pointer becomes a suitable choice as we can check for ```nil``` if the flag is not defined.
I've added code blocks like this to read default command line flag values in case of missing/wrong config file. It also checks if the command line flags override the conf, if conf file is present and correct.
	if doLookup {
		if cmd.Flags().Lookup(name).Changed {
			*f = mustGetIntFlag(cmd, name)
		}
	} else {
		*f = mustGetIntFlag(cmd, name)
	}
}
These blocks dereference the input variable pointer to assign value to it. And this is causing nil pointer exception.
I have modified the code to handle these cases and it's working fine now.
If the config file ```conf.yaml``` is not present. It picks from the default command line flag values. It also supports overriding those.

```
dynamic-buildkite-template % rm .\conf.yaml 
dynamic-buildkite-template % go run . trivy
{"level":"debug","msg":"config path:conf.yaml","time":"2023-10-02T12:51:13+01:00"}
{"level":"debug","msg":"error while loading the configuration file. Loading the defaults","time":"2023-10-02T12:51:13+01:00"}
{"level":"debug","msg":"trivy plugin configuration not found in the config file or wrong config file. Proceeding with defaults from command line flags.","time":"2023-10-02T12:51:13+01:00"}
{"level":"info","msg":"latest trivy plugin tag: v1.18.3","time":"2023-10-02T12:51:13+01:00"}
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
```

If conf.yaml is present it reads and point the values/config from the file along with zero values, also it supports command line overrides.

```
dynamic-buildkite-template % cat conf.yaml 
plugins:
  trivy:
    exit-code: 0
    timeout: 5m0s
    severity: HIGH
    ignore-unfixed: true
    security-checks: vuln,config
    skip-files: ""
    skip-dirs: ""
    image-ref: ""
    version: ""
    helm-overrides-file: ""

 dynamic-buildkite-template % go run . trivy
{"level":"debug","msg":"config path:conf.yaml","time":"2023-10-02T13:14:23+01:00"}
{"level":"info","msg":"latest trivy plugin tag: v1.18.3","time":"2023-10-02T13:14:24+01:00"}
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""

dynamic-buildkite-template % go run . trivy --severity="HIGH,CRITICAL"
{"level":"debug","msg":"config path:conf.yaml","time":"2023-10-02T13:15:16+01:00"}
{"level":"info","msg":"latest trivy plugin tag: v1.18.3","time":"2023-10-02T13:15:16+01:00"}
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
```

if a flag is not defined in the conf.yaml then you won't see that in the output.This allows the user to not define few parameters in the template.However, we can always add that using command line override. For example, I have commented the severity parameter.

```
dynamic-buildkite-template % cat conf.yaml 
plugins:
  trivy:
    exit-code: 0
    timeout: 5m0s
    #severity: HIGH
    ignore-unfixed: true
    security-checks: vuln,config
    skip-files: ""
    skip-dirs: ""
    image-ref: ""
    version: ""
    helm-overrides-file: ""
dynamic-buildkite-template % go run . trivy                           
{"level":"debug","msg":"config path:conf.yaml","time":"2023-10-02T13:20:18+01:00"}
{"level":"info","msg":"latest trivy plugin tag: v1.18.3","time":"2023-10-02T13:20:19+01:00"}
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
dynamic-buildkite-template % go run . trivy --severity="HIGH,CRITICAL"
{"level":"debug","msg":"config path:conf.yaml","time":"2023-10-02T13:20:34+01:00"}
{"level":"info","msg":"latest trivy plugin tag: v1.18.3","time":"2023-10-02T13:20:34+01:00"}
steps:
  - command: ls
    plugins:
      - equinixmetal-buildkite/trivy#v1.18.3:
          exit-code: 0
          timeout : "5m0s"
          severity: "HIGH,CRITICAL"
          ignore-unfixed: true
          security-checks: "vuln,config"
          skip-files: ""
          skip-dirs: ""
          image-ref: ""
          helm-overrides-file: ""
```

prezha commented:
/lgtm
thanks @hnadiminti-equinix for fixing this!

hnadiminti-equinix commented:
> /lgtm thanks @hnadiminti-equinix for fixing this!

Thank you @prezha !
